src/types/migration/document.ts

```diff
@@ -0,0 +1,223 @@
+ *
+ * @typeParam TField - Field type to convert.
+ */
+export type FieldToMigrationField<
+	TField extends AnyRegularField | GroupField | SliceZone,
+> = TField extends AnyRegularField
+	? RegularFieldToMigrationField<TField>
+	: TField extends GroupField
+		? GroupFieldToMigrationField<TField>
+		: TField extends SliceZone
+			? SliceZoneToMigrationField<TField>
+			: never
+
+/**
+ * A utility type that converts a record of fields to a record of migration
+ * fields.
+ *
+ * @typeParam TFields - Type of the record of Prismic fields.
+ */
+export type FieldsToMigrationFields<
+	TFields extends Record<string, AnyRegularField | GroupField | SliceZone>,
+> = {
+	[Key in keyof TFields]: FieldToMigrationField<TFields[Key]>
+}
+
+/**
+ * Makes the UID of `PrismicMigrationDocument` optional on custom types without
+ * UID. TypeScript fails to infer correct types if done with a type
+ * intersection.
+ *
+ * @internal
+ */
+type MakeUIDOptional<TMigrationDocument extends { uid: string | null }> =
+	TMigrationDocument["uid"] extends string
+		? TMigrationDocument
+		: Omit<TMigrationDocument, "uid"> & Partial<Pick<TMigrationDocument, "uid">>
+
+/**
+ * A Prismic document compatible with the migration API.
+ *
+ * @see More details on the migraiton API: {@link https://prismic.io/docs/migration-api-technical-reference}
+ */
+export type PrismicMigrationDocument<
+	TDocument extends PrismicDocument = PrismicDocument,
+> =
+	TDocument extends PrismicDocument<infer TData, infer TType, infer TLang>
+		? MakeUIDOptional<{
+				/**
+				 * Type of the document.
+				 */
+				type: TType
+
+				/**
+				 * The unique identifier for the document. Guaranteed to be unique among
+				 * all Prismic documents of the same type.
+				 */
+				uid: TDocument["uid"]
+
+				/**
+				 * Language of document.
+				 */
+				lang: TLang
+
+				/**
+				 * The identifier for the document. Used for compatibily with the
+				 * content API.
+				 *
+				 * @internal
+				 */
+				// Made optional compared to the original type.
+				id?: TDocument["id"]
+
+				/**
+				 * Alternate language documents from Prismic content API. Used as a
+				 * substitute to the `masterLanguageDocument` options when the latter is
+				 * not available.
+				 *
+				 * @internal
+				 */
+				// Made optional compared to the original type.
+				alternate_languages?: TDocument["alternate_languages"]
+
+				/**
+				 * Tags associated with document.
+				 */
+				// Made optional compared to the original type.
+				tags?: TDocument["tags"]
+
+				/**
+				 * Data contained in the document.
+				 */
+				data: FieldsToMigrationFields<TData>
+			}>
+		: never
+
+/**
+ * Parameters used when creating a Prismic document with the migration API.
+ *
+ * @see More details on the migraiton API: {@link https://prismic.io/docs/migration-api-technical-reference}
+ */
+export type PrismicMigrationDocumentParams = {
+	/**
+	 * Name of the document displayed in the editor.
+	 */
+	documentName: string
```

Comment:
I made that so that it's consistent with asset names:
```typescript
writeClient.createAsset(file, filename)
writeClient.createDocument(document, documaneName)
```

Do we want to teach users about the API itself and stay consistent with it here? No strong opinion on my end.